open-api/rest-catalog-open-api.yaml

```diff
@@ -3647,6 +3818,176 @@ components:
+      allOf:
+        - $ref: '#/components/schemas/BaseTableScanContext'
+      required:
+        - type
+        - start-snapshot-id
+      properties:
+        type:
+          type: string
+          enum: ["incremental-snapshot-scan"]
+        select:
+          $ref: '#/components/schemas/SelectedFieldNames'
+        filter:
+          $ref: '#/components/schemas/Filter'
+        case-sensitive:
+          description: If field selection and filtering should be case sensitive
+          type: boolean
+          default: true
+        start-snapshot-id:
+          description: The ID of the starting snapshot of the incremental scan
+          type: integer
+          format: int64
+        inclusive-start:
+          description: If the data appended in the start snapshot should be included in the scan
+          type: boolean
+          default: false
+        end-snapshot-id:
+          description:
+            The ID of the inclusive ending snapshot of the incremental scan.
+            If not specified, the snapshot at the main branch head will be used as the end snapshot.
+          type: integer
+          format: int64
+
+    FieldName:
+      description:
+        A field name that follows the Iceberg naming standard, and can be used in APIs like
+        Java `Schema#findField(String name)`.
+
+        The nested field name follows these rules
+        - nested struct fields are named by concatenating field names at each struct level using dot (`.`) delimiter,
+        e.g. employer.contact_info.address.zip_code
+        - nested fields in a map key are named using the keyword `key`, e.g. employee_address_map.key.first_name
+        - nested fields in a map value are named using the keyword `value`, e.g. employee_address_map.value.zip_code
+        - nested fields in a list are named using the keyword `element`, e.g. employees.element.first_name
+      type: string
+
+    SelectedFieldNames:
+      description:
+        A list of fields in schema that are selected in a table scan.
+        When not specified, all columns in the requested schema should be selected.
+      type: array
+      items:
+        $ref: '#/components/schemas/FieldName'
+
+    Filter:
+      description:
+        an unbounded expression to describe the filters to apply to a table scan, 
+        default to `TrueExpression` meaning that nothing is filtered.
+      allOf:
+        - $ref: '#/components/schemas/Expression'
+      default: { "type": "true" }
+
+    PlanTask:
+      description:
+        An opaque JSON object that contains information provided by the REST server
+        to be utilized by clients for distributed table scan planning; should be supplied
+        as input in `PlanTable` operation.
+      type: object
+
+    FileScanTask:
+      type: object
+      required:
+        - data-file
+        - residual-filter
+      properties:
+        data-file:
+          $ref: '#/components/schemas/DataFile'
+        position-delete-files:
```

Comment:
I don't understand the rationale for using separate lists of delete files and I think that this limits our ability to add more delete files in the future.

If we were to add a new delete file format, it would make sense to add another list here. But clients don't know about other properties so it would require a more invasive spec change. If instead we used an array of `DeleteFile` that is a `oneOf` with `PositionDeleteFile` and `EqualityDeleteFile` then clients would not be able to continue if they encounter a different `content` type in `ContentFile`.

Reply:
Separate from the comment above, we may want to change this to a list of references to delete files instead of a list of the files themselves. There are many cases where the same delete files can be used for multiple data files. With the current structure, a delete file that is applied to every data file would be copied into every FileScanTask sent from the service. Instead, we could add a list of delete files and a list of file scan tasks and reference the delete files by an ID or by position in the top-level list.

An added benefit of using a separate delete file list is that it is more flexible. For example, an engine may perform a scan by reading delete files and data files separately and performing an anti-join. Not needing to pre-process and deduplicate the delete files would be better for that case.

Reply:
Thanks for the insights on this. So to confirm you basically are saying that you would like the list of actual deletes files to exist on the same level as the list of file scan tasks. Something like this?
```
    PlanTableResult:
      type: object
      required:
        - file-scan-tasks
      properties:
        file-scan-tasks:
          type: array
          items:
            $ref: '#/components/schemas/FileScanTask'
        delete-files:
          type: array
          items:
              $ref: '#/components/schemas/DeleteFile'
```
And then within the the `FileScanTask` there will be a list of ID/positions mapping to those actual delete files, something like this:
```
    FileScanTask:
      type: object
      required:
        - data-file
        - residual-filter
      properties:
        data-file:
          $ref: '#/components/schemas/DataFile'
        delete-files-references:
          description: A list of delete file IDs
          type: array
          items:
            type: integer

```

Reply:
> a client can choose whether or not it can handle that new delete file type.

One small callout: we should make sure that the client is *required* to fail if it does not support or recognize a delete file type and that delete file is used by a FileScanTask (or DataFile if that's what we use instead).

Reply:
Ah yes, bad wording on my part to say "can choose", to ensure correctness clients should be required to fail if they hit a delete file they cannot support or recognize.

Reply:
+1 for indices